### DISCUSSION TITLE
Are Contexts always able to be evaluated in a runner script?

### DISCUSSION THREAD

**Discussion by oblivioncth** [community/community#26720] [Actions]
<p>This is likely because of the rapid evolution of GitHub Actions/Workflows, but when referencing many examples/tutorials (and some GitHub Actions docs) on the web, many note that when using environment variables you have to handle expansion using the appropriate syntax for your given shell</p>
<p>For example:</p>
<pre><code class="lang-yaml">job:
  print-stuff:
    runs-on: *windows/linux, depending on shell usage*
    env:
      some_var: Hello World
    steps:
    - shell: pwsh
      run: echo $Env:some_var
    - shell: cmd
      run: echo %some_var%
    - shell: bash
      run: echo $some_var
</code></pre>
<p>This makes sense since the runner is actually processing these lines directly, especially for variables you set within the runner itself and not in the workflow file.</p>
<p>But then, in a couple places, including the <a href="https://docs.github.com/en/enterprise-server@3.2/actions/learn-github-actions/contexts#example-usage-of-the-env-context" rel="noopener nofollow ugc">official docs</a>, I’ve seen syntax like this:</p>
<pre><code class="lang-yaml">job:
  print-stuff:
    runs-on: *whatever*
    env:
      some_var: Hello World
    steps:
      run: echo ${{ env.some_var }}
</code></pre>
<p>Until I found this in the documentation I was confused as I thought this wasn’t allowed, given all of the statements regarding the necessity of per-shell variable expansion syntax. Perhaps this is a newer feature, since the few examples I’ve found of it are newer?</p>
<p>So basically my question is, is it always OK to resolve environment (or other context) variables at the “workflow level” in a ‘run’ statement using expressions like this? Are there any gotchyas/things to watch out for?</p>
<p>Obviously the context has to be available in a given situation and variables in the ‘env’ context are only there if defined in the workflow or written back using the environment file.</p>
<p>It seems like the answer is yes, but again I’m just a bit unsure now given all of the talk about shell specific variable expansion.</p>
<p>This just seems much easier to do when availability permits since its shell agonistic.</p>

**Comment by airtower-luna** [comment]
<aside class="quote no-group" data-username="oblivioncth" data-post="1" data-topic="245017">
<div class="title">
<div class="quote-controls"></div>
 oblivioncth:</div>
<blockquote>
<p>So basically my question is, is it always OK to resolve environment (or other context) variables at the “workflow level” in a ‘run’ statement using expressions like this? Are there any gotchyas/things to watch out for?</p>
</blockquote>
</aside>
<p>The main thing to keep in mind is that <code>${{ ... }}</code> is Actions syntax: It has absolutely nothing to do with the shell, and the shell never knows about it. It’s templating, basically: The Actions runner substitutes the <code>${{ ... }}</code> with whatever the result of the expression between the double curly brackets is to make a script file. That script file is then passed to the shell for the <code>run</code> step. So if the result of the expression contains any characters the shell treats specially (e.g. spaces, quotes, semicolons, …), you may get odd to dangerous results.</p>
<p>When you use environment variables with shell syntax you get whatever processing/escaping your shell of choice does.</p>
<p>Of course you need to be careful with untrusted input either way, and validate or escape it as appropriate.</p>

**Comment by oblivioncth** [comment]
<p>Ok, I understood that the expressions are evaluated at the “Actions level”, being passed to the shell with all substitutions already performed, but thanks for the notes on what to watch for when approaching variables this way as I do see how it could be a bit easier to miss special characters being passed in without the proper escapes.</p>
<p>I’m also just curious about the history of this use given all of the instances of guides implying variable substitution <strong>must</strong> be applied by the shell in a run action, whereas it sounds like, at least today, its more of a recommendation than a requirement.</p>
<p>Wasn’t sure if it used to be the only option, or is just the way that most prefer, and therefore is spoke about in a way as if it was <em>the</em> method.</p>

**Comment by airtower-luna** [comment, marked as answer]
<aside class="quote no-group" data-username="oblivioncth" data-post="3" data-topic="245017">
<div class="title">
<div class="quote-controls"></div>
 oblivioncth:</div>
<blockquote>
<p>I’m also just curious about the history of this use given all of the instances of guides implying variable substitution <strong>must</strong> be applied by the shell in a run action, whereas it sounds like, at least today, its more of a recommendation than a requirement.</p>
</blockquote>
</aside>
<p>The templating syntax has been available since public availability of Actions, I don’t know if there was anything different during private beta.</p>
<p>I guess using shell variables is more common because it’s natural to do it that way if you’re familiar with shell scripting. Just write like you would elsewhere, without switching between shell and Actions template. Also passing variables via <code>env</code> makes it nicely obvious what gets passed in, including in the log.</p>
<p>In the end I think it’s a matter of choosing the tool that best fits your needs. :slightly_smiling_face:</p>

**Comment by oblivioncth** [comment]
<p>Ah, thanks for the clarification.</p>
<p>Perhaps I was exaggerating how strict the references I’m referring to sounded, but that was still my overall initial impression.</p>
<p>With the “luxury” of being someone just getting started with this, I will be trying to use the <code>env</code> context as much as possible, so that’s nice to hear the logs will be a little more helpful in that case. Though, I’m certainly a bit irked by the fact you can’t set variables in the <code>env</code> path using other <code>env</code> context variables, and instead need to do so using the shell. Hopefully in the the workflow script parser becomes powerful enough to allow for that :sweat_smile:.</p>

**Comment by airtower-luna** [comment]
<aside class="quote group-Advisory-Council" data-username="airtower-luna" data-post="2" data-topic="245017">
<div class="title">
<div class="quote-controls"></div>
 airtower-luna:</div>
<blockquote>
<p>Of course you need to be careful with untrusted input either way, and validate or escape it as appropriate.</p>
</blockquote>
</aside>
<p>One addition: The article linked below is a good description of what can go wrong with untrusted input. And yes, it’s easier to do wrong with templating. :sweat_smile:</p>
<aside class="onebox allowlistedgeneric" data-onebox-src="https://securitylab.github.com/research/github-actions-untrusted-input/">
  <header class="source">
      <img src="https://user-images.githubusercontent.com/5814919/181099388-7e4b9080-4d3a-4715-a5e4-1c72176652a4.png" class="site-icon" width="72" height="72">

      <a href="https://securitylab.github.com/research/github-actions-untrusted-input/" target="_blank" rel="noopener" title="12:00AM - 28 January 2021">GitHub Security Lab – 28 Jan 21</a>
  </header>

  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:690/362;"><img src="https://user-images.githubusercontent.com/5814919/181099389-6d65bf58-9058-400b-ba1c-b10e202af6a1.png" class="thumbnail" width="690" height="362" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/8/4/844607e16e2775a69e8cd3cfefc2ab35ac88a928_2_690x362.png, https://aws1.discourse-cdn.com/github/optimized/3X/8/4/844607e16e2775a69e8cd3cfefc2ab35ac88a928_2_1035x543.png 1.5x, https://aws1.discourse-cdn.com/github/original/3X/8/4/844607e16e2775a69e8cd3cfefc2ab35ac88a928.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/8/4/844607e16e2775a69e8cd3cfefc2ab35ac88a928_2_10x10.png"></div>

<h3><a href="https://securitylab.github.com/research/github-actions-untrusted-input/" target="_blank" rel="noopener">Keeping your GitHub Actions and workflows secure Part 2: Untrusted input</a></h3>

  <p>In this article, we’ll discuss possible avenues of abuse that may result in code and command injection in otherwise seemingly secure workflows. Our examples are based on real-world GitHub workflow implementation vulnerabilities the GitHub Security...</p>


  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

**Comment by oblivioncth** [comment]
<p>I appreciate the reference.</p>
<p>This does explain why some just avoid using the GitHub expressions when writing run sections altogether. Otherwise I will need to pay close attention to which context variables are used, and where, with what events I trigger a given workflow, and who exactly is interacting with the repository.</p>
<p>I’ll definitely be more careful and not just blindly use ${{ X }} all over the place.</p>
